crates/metadata-providers/src/dynamodb.rs

```diff
@@ -0,0 +1,492 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use aws_config::{BehaviorVersion, Region, timeout::TimeoutConfig};
+use aws_sdk_dynamodb::{
+    config::Credentials,
+    error::SdkError,
+    operation::{delete_item::DeleteItemError, put_item::PutItemError},
+    primitives::Blob,
+    types::{AttributeValue, ReturnConsumedCapacity, ReturnValue},
+};
+use bytestring::ByteString;
+
+use restate_metadata_store::{ProvisionedMetadataStore, ReadError, WriteError};
+use restate_types::{
+    Version,
+    config::{DynamoDbOptions, MetadataClientOptions},
+    metadata::{Precondition, VersionedValue},
+    net::connect_opts::CommonClientConnectionOptions,
+};
+
+const KIND_FIELD: &str = "kind";
```

Comment:
Nitpick: wouldn't `key` be more a bit more accurate? Technically `pp_epoch_0` and `pp_epoch_1` are the same "kind" of metadata but distinct keys 🤓

Reply:
+1 for not calling it kind but maybe `key-name` or `pk` or `pkey`?